### PR TITLE
lib.modules: Let module declare options directly in bare submodule

### LINKED ISSUE #146882: Can't merge hierarchical options into submodule (or freeform option)
### Describe the bug

Part of an option tree may be free-form.

Extending the free-form part in a separate module is hard and prevents certain use cases and migrations.

### Steps To Reproduce

```nix
let
  lib = import ./lib;
  inherit (lib) types mkOption;
in
(lib.evalModules {
  modules = [
    {
      options.free = mkOption {
        type = lib.types.submoduleWith {
          modules = [
            # UPDATE: A freeformType is not necessarily part of the problem.
            # { freeformType = lib.types.anything; }
          ];
        };
        default = {};
      };
    }
    {
      options.free = mkOption {
        type = lib.types.submoduleWith {
          modules = [
            {
              options.complicated = mkOption {
                type = types.int;
                default = 1;
              };
            }
          ];
        };
      };
    }
    {
      options.free.specific = mkOption {
        type = lib.types.int;
        default = 2;
      };
    }
  ];
}).config
```

```console
$ nix-instantiate test.nix  --eval --strict
error: The option `free' in `<unknown-file>' is a prefix of options in `<unknown-file>'.
(use '--show-trace' to show detailed location information)

```

The submodule only exists in order to provide a freeform type for that subtree, but it is conceptually still the same tree.

### Expected behavior

```console
$ nix-instantiate test.nix  --eval --strict
{ free = { complicated = 1; specific = 2; }; }
```

### Additional context

A semi-workaround is to make writing the unnecessary module boundary easier.

```
    mkSubmoduleOptions =
      options:
      mkOption {
        type = types.submoduleWith {
          modules = [ { inherit options; } ];
        };
      };
```

However, this is needlessly complicated and it does not help in situations where a freeform type is introduced in a decentralized scenario.

### Notify maintainers

@Infinisil 

### Metadata

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
  - lib.evalModules
  - lib.mkOption
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by roberth**
###### Motivation for this change

Let module declare options directly in bare submodule, where a bare submodule is an option that has a type like `submoduleWith x`, as opposed to `attrsOf (submoduleWith x)`.

This makes migration unnecessary when introducing a freeform type in an existing option tree and it removes some unwieldy syntax. Refs https://github.com/NixOS/nixpkgs/pull/156503/files#r790768951

Mainly, it allows the introduction of a freeformType at any place in an existing options tree without breaking compatibility with existing option declarations inside the new freeformType.

Another use case is for modules that should be imported in multiple locations in the options tree, such as `systemd` and `systemd.user`.

Closes #146882

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: lib/modules.nix [hunk 1]
```diff
@@ -471,6 +471,18 @@ rec {
           [{ inherit (module) file; inherit value; }]
         ) configs;
 
+      # Convert an option tree decl to a submodule option decl
+      optionTreeToOption = decl:
+        if builtins.trace decl isOption decl.options
```
Comment (ncfavier) [inline lib/modules.nix:476]: Don't forget to remove the trace

File: lib/modules.nix [hunk 2]
```diff
@@ -490,7 +502,15 @@ rec {
...
               firstNonOption = findFirst (m: !isOption m.options) "" decls;
             in
-              throw "The option `${showOption loc}' in `${firstOption._file}' is a prefix of options in `${firstNonOption._file}'."
+              if firstOption.options.type?isSubmodule && firstOption.options.type.isSubmodule
+              then
+                let opt = fixupOptionType loc (mergeOptionDecls loc (map optionTreeToOption decls));
+                in {
+                  matchedOptions = evalOptionValue loc opt defns';
+                  unmatchedDefns = [];
+                }
```
Comment (ncfavier) [inline lib/modules.nix:511]: This could probably be deduplicated with the code just above.

File: lib/types.nix [hunk 3]
```diff
@@ -594,6 +594,11 @@ rec {
               else throw "A submoduleWith option is declared multiple times with conflicting shorthandOnlyDefinesConfig values";
           };
         };
+      } // {
+        # Submodule-typed options get special treatment in order to facilitate
+        # certain migrations, such as the addition of freeformTypes onto
+        # existing option trees.
+        isSubmodule = true;
```
Comment (ncfavier) [inline lib/types.nix:601]: I think for further extensibility it would make sense to abstract this into a function, so that e.g. `listOf` can define its own logic for things like `foo."*".bar`. This would replace `optionTreeToOption` (or perhaps `map optionTreeToOption`) in `lib/modules`.

File: lib/modules.nix (hunk 2 above)
Comment (roberth) [reply lib/modules.nix:520]: Done.
It could be made even less repetitive, but that would require more conditionals or function calls that I'd prefer not to add because of evaluation performance.

File: lib/types.nix (hunk 3 above)
Comment (roberth) [reply lib/types.nix]: I agree, but I can't decide this without also deciding what the semantics of, say, `attrsOf` should be when it comes to option merging. It's ok to take small steps. I've added an explanation as an inline comment.

I've removed `isSubmodule` is it wasn't strictly necessary. This way we won't have to remove it when we do generalize this feature.

File: lib/modules.nix (hunk 2 above)
Comment (ncfavier) [reply lib/modules.nix:520]: I think what I meant was that the first two cases can be combined into one as they return the same value (`map optionTreeToOption decls == decls` in the first case).

```nix
if optionDecls == [] then
  mergeModules' loc decls defns
else if (lib.head optionDecls).options.type.name != "submodule" then
  throw "The option ... is a prefix of ..."
else
  let opt = fixupOptionType loc (mergeOptionDecls loc (map optionTreeToOption decls));
  in {
    matchedOptions = evalOptionValue loc opt defns';
    unmatchedDefns = [];
  }
```

I don't think this hurts performance? It might slightly change the semantics if `decls == []`, but is that case even possible?

Comment (roberth) [reply lib/modules.nix:519]: > I guess it calls optionTreeToOption more.

Exactly. Calls are quite expensive.

> make `map optionTreeToOption` a fold

I don't expect going from two list function to a single currying list function to be an improvement. Having to then store both results as an intermediate value is very likely to make it more expensive.

> Does this kind of micro-optimisation really matter in practice?

Often it doesn't, but this code is called very frequently. OfBorg's total allocated bytes are down by 1/800th. I'd say that's not insignificant. I'd speculate based on the numbers that a `builtins.count` could achieve four times that, or 0.5%.

I don't think we need to improve performance any further in this PR, but It will be interesting to try something in a follow up, so we can see the effect of a single change.

File: lib/modules.nix [hunk 4]
```diff
@@ -475,26 +474,54 @@ rec {
...
               unmatchedDefns = [];
             }
-          else if nrOptions != 0 then
-            let
-              firstOption = findFirst (m: isOption m.options) "" decls;
-              firstNonOption = findFirst (m: !isOption m.options) "" decls;
-            in
-              throw "The option `${showOption loc}' in `${firstOption._file}' is a prefix of options in `${firstNonOption._file}'."
+          else if optionDecls != [] then
+              if (lib.head optionDecls).options.type.name == "submodule"
```
Comment (Infinisil) [inline lib/modules.nix:504]: This leads to a different error depending on module order. E.g.
```nix
let
  lib = import ./lib;
in lib.evalModules {
  modules = [
    {
      _file = "a.nix";
      options.foo = lib.mkOption {
        type = lib.types.submoduleWith { modules = []; };
      };
    }
    {
      _file = "b.nix";
      options.foo = lib.mkOption {
        type = lib.types.int;
      };
    }
    {
      _file = "c.nix";
      options.foo.bar = lib.mkOption {
        type = lib.types.str;
      };
    }
  ];
}
```

gives

```
error: The option `foo' in `b.nix' is a prefix of options in `c.nix'.
(use '--show-trace' to show detailed location information)
```

While

```nix
let
  lib = import ./lib;
in lib.evalModules {
  modules = [
    {
      _file = "b.nix";
      options.foo = lib.mkOption {
        type = lib.types.int;
      };
    }
    {
      _file = "a.nix";
      options.foo = lib.mkOption {
        type = lib.types.submoduleWith { modules = []; };
      };
    }
    {
      _file = "c.nix";
      options.foo.bar = lib.mkOption {
        type = lib.types.str;
      };
    }
  ];
}
```

gives

```
error: The option `foo' in `b.nix' is already declared in `c.nix' and `a.nix'.
(use '--show-trace' to show detailed location information)
```

Not yet sure what the best error to throw is in such a case

File: lib/modules.nix [hunk 5]
```diff
@@ -475,26 +474,54 @@ rec {
...
 
+      # Convert an option tree decl to a submodule option decl
+      optionTreeToOption = decl:
+        if isOption decl.options
+        then decl
+        else decl // {
+            options = mkOption {
+              type = types.submoduleWith {
+                modules = [ { options = decl.options; } ];
+              };
```
Comment (Infinisil) [inline lib/modules.nix:485]: This leads to an error when a users option is declared with just `types.submodule`:

```nix
let
  lib = import ./lib;
in lib.evalModules {
  modules = [
    {
      _file = "a.nix";
      options.foo = lib.mkOption {
        type = lib.types.submodule {};
        default = {};
      };
    }
    {
      _file = "b.nix";
      options.foo.bar = lib.mkOption {
        type = lib.types.str;
      };
    }
  ];
}
```

gives

```
error: A submoduleWith option is declared multiple times with conflicting shorthandOnlyDefinesConfig values
(use '--show-trace' to show detailed location information)
```

Not sure if this can be fixed nicely or if it even makes sense.

**Review by infinisil — CHANGES_REQUESTED** [review]

File: lib/modules.nix (hunk 4 above)
Comment (roberth) [reply lib/modules.nix:504]: Yeah, I've inherited this flaw from the existing code.
Might have a look later, but not sure if this needs to be scoped in.

File: lib/modules.nix (hunk 5 above)
Comment (roberth) [reply lib/modules.nix:485]: `submodule` is the legacy one, but yeah, not great. Seems like "don't care" needs to be representable and pick `true` as the default. `false` was probably a bad idea.

File: lib/modules.nix (hunk 4 above)
Comment (Infinisil) [reply lib/modules.nix:504]: Previously the order of modules only had an influence on the specific files mentioned in the error message, but the error message stayed the same. This PR would make the error message variety itself change.

**Comment by Infinisil** [comment]
I wonder if it would be feasible for option trees (like `options.foo.bar.baz = mkOption { ... }`) to just be desugared to submodules (like `options.foo = mkOption { type = submodule { options.bar = mkOption { type = submodule { options.baz = mkOption { ... }; }; }; }; }`) and in the process cleaning up the code :thinking:

File: lib/modules.nix (hunk 5 above)
Comment (roberth) [reply lib/modules.nix]: Added `null` support, but only when merging.

File: lib/modules.nix (hunk 4 above)
Comment (roberth) [reply lib/modules.nix]: It now checks all and I've improved the error message.

**Comment by roberth** [comment]
> I wonder if it would be feasible for option trees to just be desugared to submodules and in the process cleaning up the code

It occurred to me too, but it wasn't clear to me whether it will be effective.
Calling back to "high-level" functions like `submodule` may not be great for performance, so it seems like it'd be a significant rethink. Not something I'll explore in this PR.

File: lib/types.nix [hunk 6]
```diff
@@ -608,7 +608,11 @@ rec {
               then lhs.specialArgs // rhs.specialArgs
               else throw "A submoduleWith option is declared multiple times with the same specialArgs \"${toString (attrNames intersecting)}\"";
             shorthandOnlyDefinesConfig =
-              if lhs.shorthandOnlyDefinesConfig == rhs.shorthandOnlyDefinesConfig
+              if lhs.shorthandOnlyDefinesConfig == null
+              then rhs.shorthandOnlyDefinesConfig
+              else if rhs.shorthandOnlyDefinesConfig == null
+              then lhs.shorthandOnlyDefinesConfig
+              else if lhs.shorthandOnlyDefinesConfig == rhs.shorthandOnlyDefinesConfig
```
Comment (Infinisil) [inline lib/types.nix:615]: Since `null` is now an allowed value, we should also define its semantics and document it. The semantics should be that if this value is needed to make a decision and it's `null`, an error is thrown. This means that if you do any shorthand definitions, you should get an error, unless another option declaration specifies the value of `shorthandOnlyDefinesConfig`, e.g. whereas

```nix
let
  lib = import ./lib;
in lib.evalModules {
  modules = [
    {
      options.foo = lib.mkOption {
        type = lib.types.submoduleWith {
          modules = [];
          shorthandOnlyDefinesConfig = null;
        };
      };

      config.foo.bar = 10;
    }
  ];
}
```

with this PR currently throws
```
error: value is null while a Boolean was expected

       at /home/infinisil/src/nixpkgs/lib/types.nix:538:23:

          537|           then setFunctionArgs (args: unify (value args)) (functionArgs value)
          538|           else unify (if shorthandOnlyDefinesConfig then { config = value; } else value);
             |                       ^
          539|
(use '--show-trace' to show detailed location information)
```

it should instead be something like

```
error: shorthandOnlyDefinesConfig is null
(use '--show-trace' to show detailed location information)
```

**Review by infinisil — CHANGES_REQUESTED** [review]

File: lib/types.nix (hunk 6 above)
Comment (roberth) [reply lib/types.nix:619]: > Since `null` is now an allowed value

I disagree. You'd have to read the implementation to find that null is acceptable in some situations.

To keep things simple and, I've decided to default it to `true`. I don't see much point in documenting this though.

File: lib/modules.nix [hunk 7]
```diff
@@ -474,26 +488,58 @@ rec {
...
+              # attrsets might be nice, but ambiguous. Suppose we have
+              # attrset as a `attrsOf submodule`. User declares option
+              # attrset.foo.bar, this could mean:
+              #  a. option `bar` is only available in `attrset.foo`
+              #  b. option `foo.bar` is available in all `attrset.*`
+              #  c. reject and require "<name>" as a reminder that it behaves like (b).
+              #  d. magically combine (a) and (c).
+              # All of the above are merely syntax sugar though.
+              then
+                let opt = mergeOptionDecls loc (map optionTreeToOption decls);
```
Comment (Infinisil) [inline lib/modules.nix:530]: After a lot of trial and error, trying to prove why `fixupOptionType` should be used here or not, I figured it out: It's needed for the sake of file locations in error messages. E.g. with

```nix
let lib = import ./lib;
in lib.evalModules {
  modules = [
    {
      _file = "a.nix";
      options.foo = lib.mkOption {
        type = lib.types.submodule {
          options.x = lib.mkOption {
            type = lib.types.int;
          };
        };
        default = {};
      };
    }
    {
      _file = "b.nix";
      options.foo.x = lib.mkOption {
        type = lib.types.str;
      };
    }
  ];
}
```

it currently produces this error:

```
$ nix-instantiate --eval -A config.foo.x
error: The option `foo.x' in `<unknown-file>' is already declared in `<unknown-file>'.
(use '--show-trace' to show detailed location information)
```

whereas with a `fixupOptionType` here, it's

```
error: The option `foo.x' in `a.nix' is already declared in `b.nix'.
(use '--show-trace' to show detailed location information)
```

would be good to have a test case for this

File: lib/modules.nix [hunk 8]
```diff
@@ -474,26 +488,58 @@ rec {
...
 
+      # Convert an option tree decl to a submodule option decl
+      optionTreeToOption = decl:
+        if isOption decl.options
+        then decl
+        else decl // {
+            options = mkOption {
+              type = types.submoduleWith {
+                modules = [ { options = decl.options; } ];
+                shorthandOnlyDefinesConfig = null;
```
Comment (Infinisil) [inline lib/modules.nix:499]: This should have a comment explaining the reasoning behind this: How it's a bit hacky but needed as a default, that `null` is only an internally valid value, that it's only safe here because of the merging function patch and because it's always going to be merged with at least one other module.

In the future this might become obsolete with https://github.com/NixOS/nixpkgs/issues/162398

**Review by infinisil — CHANGES_REQUESTED** [review]

File: lib/modules.nix (hunk 7 above)
Comment (roberth) [reply lib/modules.nix]: Done.

File: lib/modules.nix (hunk 8 above)
Comment (roberth) [reply lib/modules.nix:502]: Added

```
                # `null` is not intended for use by modules. It is an internal
                # value that means "whatever the user has declared elsewhere".
                # This might become obsolete with https://github.com/NixOS/nixpkgs/issues/162398
```

**Review by infinisil — APPROVED** [review]
Looking good!

**Comment by bobvanderlinden** [comment]
Could you explain the example for systemd a bit?

Would this allow something like:

```
let
  units = {
    options = {
      units = mkOption { ... };
    };
    config = { ... };
  };
in
{
  options = {
    boot.systemd = mkOption {
      type = types.submoduleWith {
        modules = [ units services mounts ... ];
      };
    };
  };
}
```

(That would be amazing)

**Comment by roberth** [comment]
@bobvanderlinden Yes, while this was already valid, it can now be mixed with "top-level" options like

```
{
  options.boot.systemd.units.foo = mkOption { };
}
```

This was not allowed to coexist with your example.
So the refactoring you want to do has become less risky and possible to do incrementally, without having to migrate all options to the submodule at once.

**Comment by roberth** [comment]
tl;dr we can improve laziness by translating bare submodules to option trees rather than the other way around, or we can implement a special "import at prefix" that does only what we need.

Submodules are more than just option trees though, so a "prefixing" operation on modules seems like it'd need quite a lot of code.
A more restricted "import at prefix" feature that only deals with options and configs may be easier to implement, as it doesn't have to translate freeformTypes. Maybe I'm overestimating the complexity of translating all submodule functionality though.

**Comment by roberth** [comment]
This PR reinterprets parts of the option tree based on information of explicitly declared options.
This could be generalized to not just the improvement of attrsets in the option tree, but also plain functions in the option tree.
Specifically, we could perhaps make the `optionType` itself responsible for interpreting the subexpressions of `options` for the purpose of static merging.

I call it static merging for two reasons:
 - It generalizes type merging. Note that types are static information. (generally! You can easily do a dependent type, but this is meant to be descriptive, not prescriptive use of the term)
 - It is static and can be used for declaring nested docs (e.g. docs in a submodule). This sets it apart from the `config` part, which is dynamic and not queried for docs.

Motivating example: https://github.com/hercules-ci/flake-parts/issues/142

**Comment by Atry** [comment]
I wonder if we could allow for merging an option with its default values, no matter what the type is.
